src/snowflake/snowpark/dataframe.py

```diff
@@ -2060,6 +2060,21 @@ def join(
                 using_columns = [using_columns]
             elif isinstance(using_columns, Column):
                 using_columns = using_columns
+            elif (
+                isinstance(using_columns, list)
+                and len(using_columns) > 0
+                and not all([type(col) == str for col in using_columns])
```

Comment:
NIT: Python probably optimizes this implicitly but I prefer `any(type(col) != str for col in using_columns)` for explicit short-circuiting.

Reply:
use isinstance, same below